mmdet3d/models/layers/pointnet_modules/voxel_sa_module.py

```diff
@@ -0,0 +1,642 @@
+            batch_num_xyzs: (Tensor): Stacked input xyz coordinates num in
+                each batch, just like (N1, N2, ...).
+            new_xyz (Tensor): Target points xyz coordinates shape with
+                (M1 + M2 ..., 3).
+            new_xyz_grid_centers (Tensor): Grids centers of each grid shape
+                with  (M1 + M2 ..., num_total_grids, 3) .
+            batch_num_new_xyzs: (Tensor): Stacked target points xyz coordinates
+                num in each batch, just like (M1, M2, ...).
+
+        Returns:
+            Tensor: Target points features shape with
+                (N1 + N2 ..., C_out).
+        """
+        with torch.no_grad():
+            if self.neighbor_type == 'ball':
+                neighbor_type = 1
+            else:
+                neighbor_type = 0
+            dist, idx, num_avg_length_of_neighbor_idxs = \
+                three_nn_vector_pool_by_two_step(
+                    support_xyz, batch_num_xyzs, new_xyz, new_xyz_grid_centers,
+                    batch_num_new_xyzs, self.max_neighbour_distance,
+                    self.nsample, neighbor_type,
+                    self.num_avg_length_of_neighbor_idxs,
+                    self.num_total_grids, self.neighbor_distance_multiplier)
+        self.num_avg_length_of_neighbor_idxs = max(
+            self.num_avg_length_of_neighbor_idxs,
+            num_avg_length_of_neighbor_idxs.item())
+
+        dist_recip = 1.0 / (dist + 1e-8)
+        norm = torch.sum(dist_recip, dim=-1, keepdim=True)
+        weight = dist_recip / torch.clamp_min(norm, min=1e-8)
+
+        empty_mask = (idx.view(-1, 3)[:, 0] == -1)
+        idx.view(-1, 3)[empty_mask] = 0
+
+        interpolated_feats = stack_three_interpolate(support_features,
+                                                     idx.view(-1, 3),
+                                                     weight.view(-1, 3))
+        interpolated_feats = interpolated_feats.view(
+            idx.shape[0], idx.shape[1],
+            -1)  # (M1 + M2 ..., num_total_grids, C)
+        if self.use_xyz:
+            near_known_xyz = support_xyz[idx.view(-1, 3).long()].view(
+                -1, 3, 3)  # ( (M1 + M2 ...)*num_total_grids, 3)
+            local_xyz = (new_xyz_grid_centers.view(-1, 1, 3) -
+                         near_known_xyz).view(-1, idx.shape[1], 9)
+            interpolated_feats = torch.cat(
+                (interpolated_feats, local_xyz),
+                dim=-1)  # ( M1 + M2 ..., num_total_grids, 9+C)
+
+        new_features = interpolated_feats.view(
+            -1, interpolated_feats.shape[-1]
+        )  # ((M1 + M2 ...) * num_total_grids, C)
+        new_features[empty_mask, :] = 0
+        if self.mlp is not None:
+            new_features = new_features.permute(
+                1, 0)[None, :, :, None]  # (1, C, N1 + N2 ..., 1)
+            new_features = self.mlp(new_features)
+
+            new_features = new_features.squeeze(dim=0).squeeze(dim=-1).permute(
```

Comment:
```suggestion
            new_features = new_features.squeeze(dim=0).squeeze(dim=-1).permute(1, 0).contiguous()
```
The `.contiguous()` in line is unnecessary in the https://github.com/open-mmlab/mmdetection3d/pull/2106/files#diff-103397effba505dde4d2de09c957e2a1c58384b335790ca14e74417eae16007cR353